repository_service_tuf_worker/repository.py

```diff
@@ -1357,92 +1373,68 @@ def sign_metadata(
+        bootstrap_state = self._settings.get_fresh("BOOTSTRAP")
+        if "signing" in bootstrap_state:
+            # Signature and threshold of initial root can only self-validate,
+            # there is no "trusted root" at bootstrap time yet.
+            if not self._validate_signature(root, signature):
+                return self._sign_result("Invalid signature", False)
+
+            root.signatures[signature.keyid] = signature
+            if not self._validate_threshold(root):
+                self.write_repository_settings("ROOT_SIGNING", root.to_dict())
+                msg = f"Root v{root.signed.version} is pending signatures"
+                return self._sign_result(msg, True)
+
+            bootstrap_task_id = bootstrap_state.split("signing-")[1]
+            self._bootstrap_finalize(root, bootstrap_task_id)
+            return self._sign_result("Bootstrap finished", True)
 
-        # Check threshold with new signature included
-        root.signatures[signature.keyid] = signature
-        if not self._validate_threshold(root):
-            self.write_repository_settings("ROOT_SIGNING", root.to_dict())
-            return self._task_result(
-                "Signature processed",
-                {
-                    "sign_metadata": True,
-                    "message": (
-                        f"Root v{root.signed.version} is pending signatures"
-                    ),
-                },
+        else:
+            # Also consult with "trusted root" when updating a new root
+            # - Signature must validate with trusted OR new root
+            # - Threshold must validate with trusted AND new root
+            trusted_root = self._storage_backend.get("root")
+            trusted_signature = self._validate_signature(
+                root, signature, trusted_root
             )
-
-        # Finalize bootstrap
-        bootstrap_task_id = bootstrap_state.split("signing-")[1]
-        self._bootstrap_finalize(root, bootstrap_task_id)
-        return self._task_result(
-            "Signature processed",
-            {
-                "sign_metadata": True,
-                "message": "Bootstrap finished",
-            },
-        )
+            new_signature = self._validate_signature(root, signature)
+
+            if not (trusted_signature or new_signature):
+                return self._sign_result("Invalid signature", False)
+
+            root.signatures[signature.keyid] = signature
+            trusted_threshold = self._validate_threshold(root, trusted_root)
+            new_threshold = self._validate_threshold(root)
+            if not (trusted_threshold and new_threshold):
+                self.write_repository_settings("ROOT_SIGNING", root.to_dict())
+                msg = f"Root v{root.signed.version} is pending signatures"
+                return self._sign_result(msg, True)
+
+            # TODO: Refactor `_root_metadata_update` to de-duplicate validation
```

Comment:
Let me know, if I should refactor `_root_metadata_update` in this PR or in a follow-up. I think it's correct as is, but performs some unnecessary cycles.

Reply:
You made a minimalist change in this PR, great work.
So, we can accommodate the refactoring here, but it is up to you.

Reply:
I suggest to do it in a follow-up. :) I can ticketize.